include/drivers/flash.h

```diff
@@ -133,6 +196,9 @@ __subsystem struct flash_driver_api {
 	flash_api_erase erase;
 	flash_api_write_protection write_protection;
 	flash_api_get_parameters get_parameters;
+	flash_api_get_page_info get_page_info;
```

Comment:
copy from: https://github.com/zephyrproject-rtos/zephyr/pull/36487#discussion_r664608864
`nvlsianpu:`
this is similar to flash-layout-get implementation api. Shouldn't it be optional?

`de-nordic:`
I was considering adding this as constant addition as the info is needed anyway, for example you can not do erase without knowing the page info, also this can be used to get alignment as it always returns info for aligned page that contains the offset.
In case of most drivers, the implementation will be quite efficient, because drivers that know about uniform layout will just do something like this:
```
if (offset < TOTAL_FLASH_SIZE) {
   fpi->offset = offset & ~(CONSTANT_PAGE_SIZE -1);
   fpi->size = CONSTANT_PAGE_SIZE);
}
```
On the other side the non uniform layout is also quick to obtain:
```
if (offset < HUGE_PAGE_OFFSET) {
   fpi->offset = offset & ~(CONSTANT_PAGE_SIZE -1);
   fpi->size = CONSTANT_PAGE_SIZE);
} else (offset < TOTAL_FLASH_SIZE) {
   fpi->offset = HUGE_PAGE_OFFSET;
   fpi->size = HUGE_PAGE_SIZE;
}
```
Because driver exactly knows the layout, in most cases, it does not have to iterate through all pages to get to the offset that is interesting for it.

Reply:
I think you are right. Option was introduced when not all devices implements the flash-get-layout API. Nowday it is almost mandatory.